python/ct/proto/client_v2.proto

```diff
@@ -0,0 +1,180 @@
+}
+
+message PrecertChainEntryV2 {
+  // pre_certificate, opaque (CMSPrecert) <1..2^24-1>
+  optional bytes pre_certificate = 1 [
+    (tls_opts).min_length = 1,
+    (tls_opts).max_length = 0xffffff ];
+
+  // precertificate_chain, repeated opaque (ASN.1Cert) <1..2^24-1>
+  // A chain from the precert to a trusted root, excluding the precert itself.
+  // Must have at least one certificate.
+  repeated bytes precertificate_chain = 2 [
+    (tls_opts).min_length = 0,
+    (tls_opts).max_length = 0xffffff,
+    (tls_opts).min_total_length = 1,
+    (tls_opts).max_total_length = 0xffffff ];
+}
+
+message LogIDV2 {
+  // DER-encoded OID without the tag byte.
+  optional bytes log_id = 1 [ (tls_opts).min_length = 2,
+                              (tls_opts).max_length = 0x7f];
+}
+
+enum SctExtensionType {
+  option (tls_enum_opts).max_value = 65535;
+
+  SCT_EXTENSION_RESERVED=65535;
+}
+
+message SctExtensionV2 {
+  optional SctExtensionType type = 1;
+  optional bytes sct_extension_data = 2 [(tls_opts).min_length = 1,
+                                      (tls_opts).max_length = 65535];
+}
+
+message SignedCertificateTimestampDataV2 {
+  optional LogIDV2 id = 1;
+  // UTC time in milliseconds, since January 1, 1970, 00:00.
+  optional uint64 timestamp = 2;
+  repeated SctExtensionV2 extensions = 3 [
+    (tls_opts).min_total_length = 0,
+    (tls_opts).max_total_length = 0xffff];
+  optional DigitallySigned signature = 4;
+}
+
+message TimestampedCertificateEntryDataV2 {
+  optional uint64 timestamp = 1;
+  optional bytes issuer_key_hash = 2 [
+    (tls_opts).min_length = 32,
+    (tls_opts).max_length = 0xff];
+  optional bytes tbs_certificate = 3 [
+    (tls_opts).min_length = 1,
+    (tls_opts).max_length = 0xffffff ];
+  repeated SctExtensionV2 sct_extensions = 4 [
+    (tls_opts).min_total_length = 0,
+    (tls_opts).max_total_length = 0xffff];
+}
+
+enum SthExtensionType {
+  option (tls_enum_opts).max_value = 65535;
+
+  STH_EXTENSION_RESERVED = 65535;
+}
+
+message SthExtension {
+  optional SthExtensionType type = 1;
+  optional bytes sth_extension_data = 2 [(tls_opts).min_length = 0,
+                                      (tls_opts).max_length = 65535];
+}
+
+message TreeHeadDataV2 {
+  optional uint64 timestamp = 1;
+  optional uint64 tree_size = 2;
+  optional bytes root_hash = 3 [
+    (tls_opts).min_length = 32,
+    (tls_opts).max_length = 0xff];
+  repeated SthExtension sth_extensions = 4 [
+    (tls_opts).min_total_length = 0,
+    (tls_opts).max_total_length = 0xffff];
+}
+
+message SignedTreeHeadDataV2 {
+  optional LogIDV2 log_id = 1;
+  optional TreeHeadDataV2 tree_head = 2;
+  optional DigitallySigned signature = 3;
+}
+
+message ConsistencyProofDataV2 {
+  optional LogIDV2 log_id = 1;
+  optional uint64 tree_size_1 = 2;
+  optional uint64 tree_size_2 = 3;
+  repeated bytes consistency_path = 4 [
+    (tls_opts).min_length = 32,
+    (tls_opts).max_length = 0xff,
+    (tls_opts).min_total_length = 32,
```

Comment:
This is intentional - no point in providing an empty consistency path (i.e., the only case where the consistency path would be empty is for a consistency proof between identical trees or between the empty tree and a non-empty tree, for both cases it makes no sense to request a consistency proof).